flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsProjectionPushDown.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.connector.source.abilities;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.table.connector.source.ScanTableSource;
+
+/**
+ * Enables to push down a (possibly nested) projection into a {@link ScanTableSource}.
+ *
+ * <p>Given the following SQL:
+ * <pre>{@code
+ *   CREATE TABLE t (i INT, r ROW < d DOUBLE, b BOOLEAN>, s STRING);
+ *   SELECT s, r.d FROM t;
+ * }</pre>
+ *
+ * <p>In the above example, {@code r.d} and {@code s} are required fields. Other fields can be skipped
+ * in a projection.
+ *
+ * <p>By default, if this interface is not implemented, a projection is applied in a subsequent operation
+ * after the source.
+ *
+ * <p>For efficiency, a source can push a projection further down in order to be close to the actual
+ * data generation. A projection is only selecting fields that are used by a query. It does neither
+ * reorder fields nor contain any computation. A projection can either be performed on the fields of
+ * the top-level row only or consider nested fields as well (see {@link #supportsNestedProjection()}).
+ *
+ * <p>Note: If a source implements {@link SupportsComputedColumnPushDown}, the projection must be applied
```

Comment:
How will that work with computed columns that are computed based on columns that could be pruned? Will those columns be pruned?

```
CREATE TABLE t (i INT, s STRING, computed AS i + 2);
SELECT computed FROM t;
```

What will be the projections? If the `i` will not be pruned does it make sense to change the order and apply projections first? Why should source bother with projecting computed columns?

Reply:
In this case, `i` will not be pruned, but `s` will be pruned. 

I think the problem is that the generated `ComputedColumnConverter` depends on the physical schema of the source. If computed column pushdown happens before projection, the pushed `ComputedColumnConverter` may not work anymore. For example, 

```
CREATE TABLE t (a STRING, i INT, s STRING, computed AS i + 2);
SELECT i,a,computed FROM t;
```

The pushed projection should be [i,a] which reordered the fields. And the pushed `ComputedColumnConverter` will get `i` from field index 0, not 1.

Reply:
Jark is right. I will extend the JavaDoc again to illustrate the problem.

Reply:
Sorry, I misread the last sentence a bit then.
> {@link SupportsComputedColumnPushDown} assumes an already applied
 projection and includes a projection of computed columns.

Just to double check. Is the order:
1. Apply projection
2. Apply computed columns?

If so what does ` includes a projection of computed columns` in that sentence mean?

Can we rephrase the last paragraph/sentence somehow? A proposal:
```
{@link SupportsComputedColumnPushDown} assumes that a projection has already been applied. The {@link ComputedColumnConverter} will use the projected physical fields and will insert computed columns into the result.
```

Reply:
I hope this is clearer now:
```
 * <p>Note: If a source implements {@link SupportsComputedColumnPushDown}, the projection must be applied
 * to the physical data in the first step. The {@link SupportsComputedColumnPushDown} (already aware of the projection)
 * will then use the projected physical data and insert computed columns into the result.
```